examples/cache.ex

```diff
@@ -1,74 +1,125 @@
+    # Pretty standard stuff, maybe apart from additionally wrapping both sides in a tuple
+    # with an extra identifier, that's just to make failure cases more obvious.
+    with {{user_id, _binary}, :parse_id} when is_snowflake(user_id) <-
+           {Integer.parse(message_user_id), :parse_id},
+         {{:ok, user}, :user} <-
+           {get_cached_with_fallback(user_id, &UserCache.get/1, &Api.get_user/1), :user},
+         {{:ok, %{name: channel_name}}, :channel} <-
+           {get_cached_with_fallback(channel_id, &ChannelCache.get/1, &Api.get_channel/1),
+            :channel},
+         {{:ok, %{name: guild_name}}, :guild} <-
+           {get_cached_with_fallback(guild_id, &GuildCache.get/1, &Api.get_guild/1), :guild} do
+      Api.create_message(
+        channel_id,
+        """
+        ID #{message_user_id} belongs to: #{User.full_name(user)}
+        Message sent in channel: ##{channel_name}
+        Channel belongs to guild: #{guild_name}
+        """
+      )
+    else
+      # Failure case for parsing and (somewhat) validating the User ID.
+      # Pretty easy to tell which point of failure this case handles,
+      # if we would match only on the return value of `Integer.parse/2`,
+      # the error value for that function is literally just `:error`,
+      # not very descriptive, is it... would not be easy to tell from a glance
+      # where in the happy path a failure occurred, if you do not know the error
+      # return values of the happy path functions in that very moment you were reading this.
+      # This kind of pattern gets pretty handy when return values are quite descriptionless or
+      # you have multiple functions that return the same thing, but also need to be handled differently.
+      # Either way, a pattern I do not see enough of, but would like to, where it's needed.
+      {_invalid_id, :parse_id} ->
+        Api.create_message(channel_id, "Make sure you entered a valid User ID")
+
+      # The cache + API failure case with the additional resource identifier, pretty straight forward,
+      # resource type for clarity, and then whatever the API throws at us as the reason, though the
+      # stringified API reason is for development purposes, not something you would throw at a regular
+      # user, but you know, this is just an example after all, make it all nice and pretty.
+      {{:error, reason}, resource} ->
+        Api.create_message(channel_id, "Unable to retrieve #{resource}: #{inspect(reason)}")
+    end
   end
+
+  # Just like handling events, a fall through `command` case to catch and ignore unwanted cases,
+  # otherwise it will go boom, KABLAM, "OH GOD NO!", *insert mushroom cloud here*, "AAAAH, MY FACE,
+  # IT'S MELTING!", well not really, since we have supervision trees, fault tolerance and the absolute,
+  # massive powerhouse of awesomeness that the BEAM is and all that jazz, but you know, you catch my drift.
+  # Also, "noop", *giggles*.
```

Comment:
I actually really enjoy the lighthearted tone throughout this example, but I think this paragraph overall might just be a *tad* too much. 😉